### PR TITLE
authorize: use opa for policy engine

### PR DESCRIPTION
**Pull request by desimone**
## Summary

These changes replace the current home-grown access policy engine with opa. The end user interface should be identical, and all the policy authorization unit tests are unchanged. These changes open the door for more complex, dynamic access policy for those who want or need it by dropping into rego. 

- authorize: policy decisions are now made based on the raw http request and unparsed jwt token
- authorize: double check jwt signature and expiry
- proxy: the raw jwt token has been added to the request context 

## Future
- Add options to allow standard policy overriding. Do we want to allow people to point at a directory of modules / rego files? 
- Add deny messages for feedback as to _why_ a particular [request was denied](https://play.openpolicyagent.org/p/9rgaoENcr3) (e.g. jwt did not verify) 
- Add rego [standalone unit tests](https://www.openpolicyagent.org/docs/latest/policy-testing/)
- Add [metrics](https://github.com/open-policy-agent/opa/blob/12ff9c0a661e3b99579349f144a217ff37150b16/server/server.go#L1013)? 

## Related issues
- #378 

**Checklist**:
- [x] add related issues
- [x] updated unit tests
- [x] ready for review

### CODE REVIEW

File: authorize/evaluator/opa/rego.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
+package opa
+
+//todo(bdd): embed source files directly, and setup tests.
```
[the remark this answers is not in the document]
Comment (desimone) [reply authorize/evaluator/opa/rego.go]: Replace or fork statik to support multiple file bundle instances. Ideally without using global state and init hooks.

File: config/policy.go [hunk 2]
```diff
@@ -16,12 +17,12 @@ type Policy struct {
 	From string `mapstructure:"from" yaml:"from"`
 	To   string `mapstructure:"to" yaml:"to"`
 	// Identity related policy
-	AllowedEmails  []string `mapstructure:"allowed_users" yaml:"allowed_users,omitempty"`
-	AllowedGroups  []string `mapstructure:"allowed_groups" yaml:"allowed_groups,omitempty"`
-	AllowedDomains []string `mapstructure:"allowed_domains" yaml:"allowed_domains,omitempty"`
+	AllowedEmails  []string `mapstructure:"allowed_users" yaml:"allowed_users,omitempty" json:"allowed_emails,omitempty"`
```
[the remark this answers is not in the document]
Comment (desimone) [reply config/policy.go]: This was completely by accident. You are right. This should be users.

File: authorize/evaluator/opa/rego.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (desimone) [reply authorize/evaluator/opa/rego.go]: `add tag to struct fields` got me in trouble here. Thoughts on renaming the Exported name to be consistent?

[the remark this answers is not in the document]
Comment (travisgroth) [reply authorize/evaluator/opa/rego.go]: >Replace or fork statik to support multiple file bundle instances. Ideally without using global state and init hooks.

Oof.  I had a feeling that was the answer :(

File: config/policy.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (travisgroth) [reply config/policy.go]: 👍

File: authorize/evaluator/opa/rego.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (travisgroth) [reply authorize/evaluator/opa/rego.go]: >add tag to struct fields got me in trouble here. Thoughts on renaming the Exported name AllowedEmails --> AllowedUsers to be consistent with the tags/policy definitions?

Yes, I think that's a good idea.  I didn't even realize the field was named that way until now.

File: config/policy.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (desimone) [reply config/policy.go]: This comment was mis-tree'd. https://github.com/pomerium/pomerium/pull/474#discussion_r373791811

^ was in response to to this request change

File: authorize/evaluator/opa/rego.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (desimone) [reply authorize/evaluator/opa/rego.go]: 👍 
> Replace or fork statik to support multiple file bundle instances. Ideally without using global state and init hooks.

I have an idea for this I can try to address in FUP PR?

File: internal/sessions/cache/cache_store.go [hunk 3]
```diff
@@ -51,26 +51,26 @@ func NewStore(o *Options) *Store {
 
 // LoadSession looks for a preset query parameter in the request body
 //  representing the key to lookup from the cache.
-func (s *Store) LoadSession(r *http.Request) (*sessions.State, error) {
+func (s *Store) LoadSession(r *http.Request) (*sessions.State, string, error) {
```
[the remark this answers is not in the document]
Comment (desimone) [reply internal/sessions/cache/cache_store.go]: I tried, it actually became a big tangled mess that was harder to reason about.

File: authorize/evaluator/opa/rego.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (travisgroth) [reply authorize/evaluator/opa/rego.go]: Yes, totally.  Was more curious than anything else.

**Pull request merged**